crates/node/ethstats/src/ethstats.rs

```diff
@@ -0,0 +1,828 @@
+            tx_root: header.transactions_root(),
+            root: header.state_root(),
+            uncles: UncleStats(vec![]),
+        })
+    }
+
+    /// Report historical block data to the `EthStats` server
+    ///
+    /// Fetches multiple blocks by their numbers and sends their statistics
+    /// to the server. This is typically called in response to a history
+    /// request from the server.
+    ///
+    /// # Arguments
+    /// * `list` - Vector of block numbers to fetch and report
+    async fn report_history(&self, list: Option<&Vec<u64>>) -> Result<(), EthStatsError> {
+        let conn = self.conn.read().await;
+        let conn = conn.as_ref().ok_or(EthStatsError::NotConnected)?;
+
+        let mut indexes = Vec::with_capacity(HISTORY_UPDATE_RANGE as usize);
+        if let Some(list) = list {
+            indexes.extend(list);
+        } else {
+            let best_block_number = self
+                .provider
+                .best_block_number()
+                .map_err(|e| EthStatsError::DataFetchError(e.to_string()))?;
+
+            let start = best_block_number.saturating_sub(HISTORY_UPDATE_RANGE);
+
+            indexes.extend(start..=best_block_number);
+        }
+
+        let mut blocks = Vec::with_capacity(indexes.size());
+        for block_number in indexes {
+            match self.provider.block_by_id(block_number.into()) {
+                Ok(Some(block)) => {
+                    blocks.push(block);
+                }
+                Ok(None) => {
+                    // Block not found, stop fetching
+                    tracing::warn!("Block {} not found", block_number);
+                    break;
+                }
+                Err(e) => {
+                    tracing::error!("Error fetching block {}: {}", block_number, e);
+                    break;
+                }
+            }
+        }
+
+        let history: Vec<BlockStats> =
+            blocks.iter().map(|block| self.block_to_stats(block)).collect::<Result<_, _>>()?;
+
+        if history.is_empty() {
+            tracing::warn!("No history to send to stats server");
+        } else {
+            tracing::debug!(
+                "Sending historical blocks to ethstats, first: {}, last: {}",
+                history.first().unwrap().number,
+                history.last().unwrap().number
+            );
+        }
+
+        let history_msg = HistoryMsg { id: self.credentials.node_id.clone(), history };
+
+        let message = history_msg.generate_history_message();
+        conn.write_json(&message).await?;
+
+        Ok(())
+    }
+
+    /// Send a complete status report to the `EthStats` server
+    ///
+    /// Performs all regular reporting tasks: ping, block info, pending
+    /// transactions, and general statistics.
+    async fn report(&self) -> Result<(), EthStatsError> {
+        self.send_ping().await?;
+        self.report_block(None).await?;
+        self.report_pending().await?;
+        self.report_stats().await?;
```

Comment:
I think this is fine but a better pattern would be something that creates the messages and then we write them all ot once so that we only need to acquire the lock once